src/Configuration.ts

```diff
@@ -0,0 +1,491 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Yargs = require("yargs");
+import * as path from "path";
+import shell = require("shelljs");
+
+export let CONFIG: ArgumentsObject;
+export class Configuration {
+  initializeConfigSingleton<T extends ArgumentsObject>(argumentValues: T) {
+    if (CONFIG) {
+      throw Error("Already initialized the config singleton!");
+    }
+
+    CONFIG = argumentValues;
+  }
+
+  private programName: string;
+
+  constructor(programName: string) {
+    this.programName = programName;
+  }
+
+  loadConfigurationFile(cwd?: string) {
+    cwd = cwd || process.env.SYNTEST_CWD || process.cwd();
+    const configPath = path.join(cwd, ".syntest.js");
+    let config;
+    // Catch syntestjs syntax errors
+    if (shell.test("-e", configPath)) {
+      try {
+        config = require(configPath);
+      } catch (error) {
+        throw new Error(error);
+      }
+      // Config is optional
+    } else {
+      config = {};
+    }
+
+    return config;
+  }
+
+  processArguments<O extends OptionsObject>(
+    yargs: O,
+    args: string[]
+  ): ArgumentsObject {
+    const config = this.loadConfigurationFile();
+    const configuredArgs = yargs.config(config);
+    return configuredArgs.wrap(configuredArgs.terminalWidth()).parseSync(args);
+  }
+
+  configureOptions() {
+    const yargs = Yargs.usage(`Usage: ${this.programName} [options]`)
+      .example(
+        `${this.programName} -r ./src`,
+        "Running the tool with target directory 'src'"
+      )
+      .example(
+        `${this.programName} -r ./src --population_size 10`,
+        "Setting the population size"
+      )
+      .epilog("visit https://syntest.org for more documentation");
+
+    const generalOptions = this.configureGeneralOptions(yargs);
+    const fileOptions = this.configureFileOptions(generalOptions);
+    const dirOptions = this.configureDirectoryOptions(fileOptions);
+    const algorithmOptions = this.configureAlgorithmOptions(dirOptions);
+    const budgetOptions = this.configureBudgetOptions(algorithmOptions);
+    const loggingOptions = this.configureLoggingOptions(budgetOptions);
+    const postProcessingOptions =
+      this.configurePostProcessingOptions(loggingOptions);
+    const samplingOptions = this.configureSamplingOptions(
+      postProcessingOptions
+    );
+    const researchModeOptions =
+      this.configureResearchModeOptions(samplingOptions);
+
+    return researchModeOptions.showHidden(false);
+  }
+
+  configureGeneralOptions<T>(yargs: Yargs.Argv<T>) {
+    return (
+      yargs
+        // plugins
+        .option("plugins", {
+          alias: ["p"],
+          default: [],
```

Comment:
We were also thinking about preparsing the plugins to put them as choices in the config. Would that already be possible in this PR or is something else required?

Reply:
The preparsing works different, you define which plugins you want to use in your config file and these then get loaded.
In PR #171 this is implemented. Since we read them from the config anyway, adding them as choices would not provide any additional benefit.

Reply:
Clear! Thank you for the clarification